### PR TITLE
add new resource util plot type

### PR DESCRIPTION
**Pull request by andre-merzky**
Hey @mturilli : this is the code for the new utilization plot.  It depends the RP PR radical-cybertools/radical.pilot/pull/2301.  Happy to iterate over it obviously, let me know what you think!

### CODE REVIEW

**Comment by mturilli** [comment]
Hi @andre-merzky , thanks that looks great. I reviewed the related RP PR and looked at this one. As we agreed before, I think it would be important to move away from a 'script' approach to RA. That is not the way most of our users need to use it and I feel we should not push for it further. I would break down the script of the PR into: a set of helper functions, some dedicated data preparation and plotting. I would then create a `src/radical/analytics/utils` and put the helper functions there. In this way, most of the code we now put into scripts will be put into functions, independent on how one likes to do plotting. Would that work for you? If so, happy to do the needful.

**Comment by andre-merzky** [comment]
Yep, I agree with that!    

> Hi @andre-merzky , thanks that looks great. I reviewed the related RP PR and looked at this one. As we agreed before, I think it would be important to move away from a 'script' approach to RA. That is not the way most of our users need to use it and I feel we should not push for it further. I would break down the script of the PR into: a set of helper functions, some dedicated data preparation and plotting. I would then create a `src/radical/analytics/utils` and put the helper functions there. In this way, most of the code we now put into scripts will be put into functions, independent on how one likes to do plotting. Would that work for you? If so, happy to do the needful.

Yup, agree with all!  Feel free to take a stab at that, but also feel free to tell me to do that! :-)

**Comment by andre-merzky** [comment]
This is ready for another round of review - thanks.

File: bin/radical-analytics-plot.py [hunk 1]
```diff
@@ -1,32 +1,100 @@
...
+    ax  = fig.add_subplot(111)
+
+    ax.spines["top"].set_visible(False)
+    ax.spines["right"].set_visible(False)
 
-plt.rc('font', **font)
+    ax.get_xaxis().tick_bottom()
+    ax.get_yaxis().tick_left()
+
+    return fig, ax
```
Comment (mturilli) [inline bin/radical-analytics-plot.py:91]: I guess all this 13--91 is now outdated and replaced by radical_mpl.txt and ra.utils.plot?

File: bin/radical-analytics-plot.py [hunk 2]
```diff
@@ -220,10 +302,11 @@ def get_elems(line):
 
 # ------------------------------------------------------------------------------
 # plot data
-plt.figure(figsize=SIZE)
 # pprint.pprint(data)
 try:
-    cnum = 0
+
+    fig, ax = fig_setup()
```
Comment (mturilli) [inline bin/radical-analytics-plot.py:308]: Should this be replaced by something that calls `get_plotsize()`?

File: bin/rp_inspect/plot_util_2.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+plt.rcParams['axes.titlesize']   = 14
+plt.rcParams['axes.labelsize']   = 14
+plt.rcParams['axes.linewidth']   =  2
+plt.rcParams['xtick.labelsize']  = 14
+plt.rcParams['ytick.labelsize']  = 14
+plt.rcParams['lines.markersize'] = 14
+plt.rcParams['lines.linewidth']  =  2
+plt.rcParams['lines.color']      = 'r'
+
+plt.rc('font', **font)
```
Comment (mturilli) [inline bin/rp_inspect/plot_util_2.py:59]: Lines 46--59 should probably be eliminated in favor of `radical_mpl.txt`

File: bin/rp_inspect/plot_util_2.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+          ]
+
+r_trans = [
+           # [{1: 'app_start'}             , 'exec_worker', 'workload'   ],
+           # [{1: 'app_stop'}              , 'workload'   , 'exec_worker'],
+           # [{1: 'dock_start'}            , 'exec_worker', 'workload'   ],
+           # [{1: 'dock_stop'}             , 'workload'   , 'exec_worker'],
+             [{1: 'req_start'}             , 'exec_worker', 'workload'   ],
+             [{1: 'req_stop'}              , 'workload'   , 'exec_worker'],
+          ]
```
Comment (mturilli) [inline bin/rp_inspect/plot_util_2.py:107]: Is this somehow duplicating the work that we have done in `rp.utils.prof_utils`?

**Review by mturilli — COMMENTED** [review]
Left some comments. I would start by discussing those before going in and try to convert all this in a set of helper functions and standalone plotting code that uses those functions. I would put those functional in `src/radical/analytics/utils/plot.py`.

File: src/radical/analytics/experiment.py [hunk 5]
```diff
@@ -45,7 +45,7 @@ def sids(self):
 
     # --------------------------------------------------------------------------
     #
-    def utilization(self, metrics, udurations=None):
+    def utilization(self, metrics, rtype='cpu', udurations=None):
```
Comment (andre-merzky) [inline src/radical/analytics/experiment.py:48]: Moving the new `rtype` argument to the end of the arg list would keep the API backward compatible.  I am not sure if that is really a concern, but I wanted to mention it.

File: src/radical/analytics/session.py [hunk 6]
```diff
@@ -979,16 +979,16 @@ def rate(self, state=None, event=None, time=None, sampling=None,
 
     # --------------------------------------------------------------------------
     #
-    def utilization(self, metrics, udurations=None):
+    def utilization(self, metrics, rtype='cpu', udurations=None):
```
Comment (andre-merzky) [inline src/radical/analytics/session.py:982]: same

File: src/radical/analytics/utils/__init__.py [hunk 7]
```diff
@@ -1,8 +1,9 @@
 # ------------------------------------------------------------------------------
 #
-from .plot import get_plotsize, get_mplstyle
+from .plot import get_plotsize, get_mplstyle, stack_transitions, get_pilot_series, get_plot_utilization, get_pilots_zeros
```
Comment (andre-merzky) [inline src/radical/analytics/utils/__init__.py:3]: line it too long

File: src/radical/analytics/utils/plot.py [hunk 8]
```diff
@@ -51,3 +56,254 @@ def get_mplstyle(name):
...
+        for p in prev:
+            stacked[m] += merged[p]
+        prev.append(m)
+
+    return stacked
+
+
+# ------------------------------------------------------------------------------
+#
+def get_pilot_series(session, pilot, tmap, resrc, use_percent=True):
```
Comment (andre-merzky) [inline src/radical/analytics/utils/plot.py:106]: a sort docstring would be great

File: src/radical/analytics/utils/plot.py [hunk 9]
```diff
@@ -51,3 +56,254 @@ def get_mplstyle(name):
...
+                    series[r][m].append([c[0], 0])
+
+    x = {'min': x_min, 'max': x_max}
+
+    return p_resrc, series, x
+
+
+# ------------------------------------------------------------------------------
+#
+def get_pilots_zeros(ra_exp_obj):
```
Comment (andre-merzky) [inline src/radical/analytics/utils/plot.py:239]: same

File: src/radical/analytics/utils/plot.py [hunk 10]
```diff
@@ -51,3 +56,254 @@ def get_mplstyle(name):
...
+            p_zeros[session.uid][pilot.uid] = pilot.timestamps(
+                event={ru.EVENT: 'bootstrap_0_start'})[0]
+
+    return p_zeros
+
+
+
+# ------------------------------------------------------------------------------
+#
+def get_plot_utilization(metrics, consumed, p_zeros, sid, pid):
```
Comment (andre-merzky) [inline src/radical/analytics/utils/plot.py:254]: same

**Review by andre-merzky — COMMENTED** [review]
This all looks great to me - thanks!  Let me run some tests with the code and apply to the `bin/rp_inspect/` scripts for consistency before merging, please.  I left a couple of comments, none being a showstopper.

Please consider this approved for your parts (I can't approve really as I am PR author...)

File: bin/rp_inspect/plot_util_2.py (hunk 3 above)
Comment (andre-merzky) [reply bin/rp_inspect/plot_util_2.py:59]: yes, will do!

File: src/radical/analytics/utils/plot.py (hunk 9 above)
Comment (mturilli) [reply src/radical/analytics/utils/plot.py:302]: Done.

File: src/radical/analytics/utils/plot.py (hunk 10 above)
Comment (mturilli) [reply src/radical/analytics/utils/plot.py:332]: Done.

File: bin/rp_inspect/plot_conc.py [hunk 11]
```diff
@@ -58,7 +61,7 @@
     src = sys.argv[1]
     session = ra.Session.create(src, 'radical.pilot')
 
-    data = {metric: session.concurrency(event=metrics[metric])
+    data = {metric: session.concurrency(event=metrics[metric], sampling=10.0)
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply bin/rp_inspect/plot_conc.py:64]: Yes, they look fine with 1s.

File: docs/source/inspection.rst [hunk 12]
```diff
@@ -3,4 +3,99 @@
...
+
+Information commonly used when analyzing and plotting one or more sessions of RADICAL-Pilot
+
+Single Session
+^^^^^^^^^^^^^^
+
+.. code-block:: python
+   :linenos:
+
+    # RTC session, pilot and task objects
```
Comment (andre-merzky) [inline docs/source/inspection.rst:19]: A typo snuck in (RTC -> RCT)

File: src/radical/analytics/experiment.py (hunk 5 above)
Comment (andre-merzky) [reply src/radical/analytics/experiment.py:48]: No, I don't think it's important, and I think we are the only users of the API.  Just want to add though that the interpreter should accept 
```py
def utilization(self, metrics, udurations=None, rtype='cpu'):
   ,,,,
```
That is perfectly valid code.

File: docs/source/inspection.rst (hunk 12 above)
Comment (mturilli) [reply docs/source/inspection.rst:19]: Done.

File: src/radical/analytics/utils/plot.py (hunk 8 above)
Comment (mturilli) [reply src/radical/analytics/utils/plot.py]: Done

File: docs/source/duration.rst [hunk 13]
```diff
@@ -82,7 +83,7 @@ To measure that duration, first, we filter the session object so to keep only th
     duration = pilots.duration(event=rp.utils.PILOT_DURATIONS_DEBUG['p_agent_runtime'])
     print(duration)
 
-.. note: This works for a set of pilots, including the corner case in which we have a single pilot. If we have a single pilot, the cumulative active time of all the pilots is equal to the active time of the only available pilot.
+.. note:: This works for a set of pilots, including the corner case in which we have a single pilot. If we have a single pilot, the cumulative active time of all the pilots is equal to the active time of the only available pilot.
```
[the remark this answers is not in the document]
Comment (mturilli) [reply docs/source/duration.rst]: Done.

File: bin/rp_inspect/plot_util_2.py (hunk 4 above)
Comment (mturilli) [reply bin/rp_inspect/plot_util_2.py:107]: Opened a discussion at https://github.com/radical-cybertools/radical.analytics/discussions/145

**Pull request merged**